File: FE/admin_app/lib/features/master_data/vehicle/data/models/vehicle_model.dart
import 'package:dart_mappable/dart_mappable.dart';

import '../../domain/entities/vehicle_entity.dart';

part 'vehicle_model.mapper.dart';

@MappableClass()
class VehicleModel extends VehicleEntity with VehicleModelMappable {
  static const List<String> headers = [
    'id',
    'plateNumber',
    'yearModel',
    'maker',
    'category',
    'orNumber',
    'crNumber',
    'fuelLevel',
    'currentMileage',
    'minDiesel',
    'withLoadConsumption',
    'withoutLoadConsumption',
    'status',
  ];

  const VehicleModel({
    super.id,
    super.plateNumber,
    super.yearModel,
    super.maker,
    super.category,
    super.orNumber,
    super.crNumber,
    super.fuelLevel,
    super.currentMileage,
    super.minDiesel,
    super.withLoadConsumption,
    super.withoutLoadConsumption,
    super.status,
  });
}

File: BE/app/models/vehicle_model.py
from sqlmodel import Field, SQLModel
from .base_model import BaseCreatedModel, BaseUpdatedModel, BaseIdModel


class VehicleBase(SQLModel):
    plateNumber: str = Field(unique=True)
    yearModel: int
    maker: str
    category: str
    orNumber: str | None
    crNumber: str | None
    fuelLevel: float = Field(default=0.00, description="In liters")
    currentMiliage: int = Field(default=0)
    minDiesel: float = Field(default=0.00)
    withLoadConsumption: float = Field(
        default=0.00, description="It should be estimated consumed km/l"
    )
    withOutLoadConsumption: float = Field(
        default=0.00, description="It should be estimated consumed km/l"
    )
    status: str = Field(description="available, unavailable, inMaintenance, retired")


class TblVehicle(
    VehicleBase, BaseCreatedModel, BaseUpdatedModel, BaseIdModel, table=True
):
    pass


class VehicleLogNoteBase(SQLModel):
    id: int = Field(default=None, primary_key=True)
    vehicleId: int = Field(foreign_key="TblVehicle.id")
    note: str


class TblVehicleLogNote(
    BaseCreatedModel,
    BaseUpdatedModel,
    VehicleLogNoteBase,
    BaseIdModel,
    table=True,
):
    pass
